src/libraries/Microsoft.Win32.SystemEvents/tests/ShutdownTest.cs

```diff
@@ -2,32 +2,45 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System;
-using System.Threading;
 using Microsoft.DotNet.RemoteExecutor;
 using Xunit;
 using static Interop;
 
-namespace Microsoft.Win32.SystemEventsTests
+namespace Microsoft.Win32.SystemEventsTests;
+
+public class ShutdownTest : SystemEventsTest
 {
-    public abstract class ShutdownTest : SystemEventsTest
+    [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNotWindowsNanoNorServerCore))]
+    [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework)]
+    public void ShutdownThroughRestartManager()
     {
-        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNotWindowsNanoNorServerCore))]
-        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework)]
-        public void ShutdownThroughRestartManager()
+        RemoteExecutor.Invoke(() =>
         {
-            RemoteExecutor.Invoke(() =>
-            {
-                // Register any event to ensure that SystemEvents get initialized
-                SystemEvents.TimeChanged += (o, e) => { };
+            // Register any event to ensure that SystemEvents get initialized
+            SystemEvents.TimeChanged += (o, e) => { };
+
+            // Fake Restart Manager behavior by sending external WM_CLOSE message
+            SendMessage(Interop.User32.WM_CLOSE, IntPtr.Zero, IntPtr.Zero);
+
+            // Emulate calling the Shutdown event
+            var shutdownMethod = typeof(SystemEvents).GetMethod("Shutdown", System.Reflection.BindingFlags.Static | System.Reflection.BindingFlags.NonPublic, null, new Type[0], null);
+            Assert.NotNull(shutdownMethod);
+            shutdownMethod.Invoke(null, null);
+        }).Dispose();
+    }
 
-                // Fake Restart Manager behavior by sending external WM_CLOSE message
-                SendMessage(Interop.User32.WM_CLOSE, IntPtr.Zero, IntPtr.Zero);
+    [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNotWindowsNanoNorServerCore))]
+    [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework)]
+    public void ShutdownSuccessDespiteThreadBlock()
+    {
+        AppContext.TryGetSwitch("Switch.SystemEvents.EnableLegacySystemEventsShutdownThreadJoin", out bool isEnabled);
+        Assert.False(isEnabled);
 
-                // Emulate calling the Shutdown event
-                var shutdownMethod = typeof(SystemEvents).GetMethod("Shutdown", System.Reflection.BindingFlags.Static | System.Reflection.BindingFlags.NonPublic, null, new Type[0], null);
-                Assert.NotNull(shutdownMethod);
-                shutdownMethod.Invoke(null, null);
-            }).Dispose();
-        }
+        RemoteExecutor.Invoke(() =>
+        {
+            // Block the SystemEvents thread. Regression test for https://github.com/dotnet/winforms/issues/11944
+            SystemEvents.UserPreferenceChanged += (o, e) => { while (true) { } };
```

Comment:
This is an artificial repro to make sure shutdown can happen despite a hang occuring. We will include a regression test of the issue scenario in the winforms repo.